File: vulnerable_script.php
<?php
/**
 * Secure vulnerable_script.php
 * Purpose: Remediation for INT310 Lab 2
 * Security improvements:
 * - Input validation & sanitization (CWE-20)
 * - SQL injection protection via prepared statements
 * - Strong password hashing (password_hash, password_verify)
 * - Secure session handling
 * - Secure file upload (whitelist + size limits + storage outside webroot)
 * - Output escaping (XSS prevention)
 * - Removed insecure session_id from GET
 * - Logging hardened (no raw user input)
 */

session_start();

// --------------------------- Database Connection ---------------------------
try {
    $pdo = new PDO("mysql:host=localhost;dbname=enterprise_app;charset=utf8mb4", "admin", "SuperSecurePassword123!", [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
        PDO::ATTR_EMULATE_PREPARES => false
    ]);
} catch (PDOException $e) {
    die("Database connection error. Please try again later.");
}

// --------------------------- Helper Functions ---------------------------
function sanitize_output($s) {
    return htmlspecialchars($s ?? '', ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8');
}

function safe_filename($name) {
    return preg_replace('/[^A-Za-z0-9._-]/', '_', basename($name));
}

// --------------------------- Registration ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['register'])) {
    $username = filter_input(INPUT_POST, 'username', FILTER_SANITIZE_STRING);
    $email    = filter_input(INPUT_POST, 'email', FILTER_SANITIZE_EMAIL);
    $password = $_POST['password'];

    if (!$username || !$email || strlen($password) < 8) {
        $error = "Invalid input. Ensure username/email are valid and password ≥ 8 characters.";
    } else {
        $hashed_password = password_hash($password, PASSWORD_DEFAULT);
        $stmt = $pdo->prepare("INSERT INTO users (username, email, password) VALUES (?, ?, ?)");
        if ($stmt->execute([$username, $email, $hashed_password])) {
            session_regenerate_id(true);
            $_SESSION['user_id']  = $pdo->lastInsertId();
            $_SESSION['username'] = $username;
            $_SESSION['role']     = 'user';
            $success = "Registration successful.";
        } else {
            $error = "Registration failed.";
        }
    }
}

// --------------------------- Login ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['login'])) {
    $username = filter_input(INPUT_POST, 'username', FILTER_SANITIZE_STRING);
    $password = $_POST['password'];

    $stmt = $pdo->prepare("SELECT id, username, role, password FROM users WHERE username = ?");
    $stmt->execute([$username]);
    $user = $stmt->fetch();

    if ($user && password_verify($password, $user['password'])) {
        session_regenerate_id(true);
        $_SESSION['user_id']  = $user['id'];
        $_SESSION['username'] = $user['username'];
        $_SESSION['role']     = $user['role'];
        $success = "Welcome, " . sanitize_output($user['username']);
    } else {
        $error = "Invalid username or password.";
    }
}

// --------------------------- Profile Update ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['update_profile'])) {
    $user_id = $_SESSION['user_id'] ?? 0;
    $bio     = filter_input(INPUT_POST, 'bio', FILTER_SANITIZE_STRING);
    $website = filter_input(INPUT_POST, 'website', FILTER_SANITIZE_URL);

    if ($user_id) {
        $stmt = $pdo->prepare("UPDATE user_profiles SET bio = ?, website = ? WHERE user_id = ?");
        if ($stmt->execute([$bio, $website, $user_id])) {
            $success = "Profile updated.";
        } else {
            $error = "Profile update failed.";
        }
    } else {
        $error = "Not authenticated.";
    }
}

// --------------------------- File Upload ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_FILES['user_file'])) {
    $user_id = $_SESSION['user_id'] ?? 0;
    $file    = $_FILES['user_file'];

    if ($user_id && $file['error'] === UPLOAD_ERR_OK) {
        $allowed = ['jpg', 'jpeg', 'png', 'pdf', 'txt'];
        $ext     = strtolower(pathinfo($file['name'], PATHINFO_EXTENSION));
        if (in_array($ext, $allowed) && $file['size'] <= 5 * 1024 * 1024) {
            $new_name = bin2hex(random_bytes(8)) . "." . $ext;
            $storage  = "/opt/lampp/private_uploads/";
            if (!is_dir($storage)) mkdir($storage, 0750, true);

            if (move_uploaded_file($file['tmp_name'], $storage . $new_name)) {
                $success = "File uploaded successfully.";
                $stmt = $pdo->prepare("INSERT INTO user_files (user_id, original_name, stored_name) VALUES (?, ?, ?)");
                $stmt->execute([$user_id, safe_filename($file['name']), $new_name]);
            } else {
                $error = "File upload failed.";
            }
        } else {
            $error = "Invalid file type or size exceeded.";
        }
    } else {
        $error = "Upload error or not authenticated.";
    }
}

// --------------------------- Search ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'GET' && isset($_GET['search'])) {
    $search_term = filter_input(INPUT_GET, 'search', FILTER_SANITIZE_STRING);
    $stmt = $pdo->prepare("SELECT name, price FROM products WHERE name LIKE ?");
    $stmt->execute(["%$search_term%"]);
    $search_results = $stmt->fetchAll();
}

// --------------------------- Admin Panel ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'GET' && isset($_GET['admin_action'])) {
    if (isset($_SESSION['role']) && $_SESSION['role'] === 'admin') {
        $action = filter_input(INPUT_GET, 'admin_action', FILTER_SANITIZE_STRING);

        if ($action === 'list_users') {
            $users = $pdo->query("SELECT id, username, email, role FROM users")->fetchAll();
        }

        if ($action === 'view_logs') {
            $log_file = basename($_GET['log_file'] ?? 'app.log');
            $log_path = "/opt/lampp/logs/app/" . $log_file;
            $log_content = file_exists($log_path) ? file_get_contents($log_path) : "Log not found.";
        }
    }
}

// --------------------------- Password Reset ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['reset_password'])) {
    $email = filter_input(INPUT_POST, 'email', FILTER_SANITIZE_EMAIL);
    if ($email) {
        $token   = bin2hex(random_bytes(16));
        $expires = date('Y-m-d H:i:s', time() + 3600);
        $stmt = $pdo->prepare("UPDATE users SET reset_token = ?, reset_expires = ? WHERE email = ?");
        $stmt->execute([$token, $expires, $email]);
        // In real system, email would be sent — here log instead
        error_log("Password reset link: https://example.com/reset?token=$token&email=$email");
        $success = "If the email exists, a reset link was sent.";
    }
}

// --------------------------- XML Processing ---------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['xml_data'])) {
    $xml_data = $_POST['xml_data'];
    $dom = new DOMDocument();
    $ok = @$dom->loadXML($xml_data, LIBXML_NONET | LIBXML_NOERROR | LIBXML_NOWARNING);
    $xml_result = $ok ? simplexml_import_dom($dom) : null;
}

// --------------------------- Logout ---------------------------
if (isset($_GET['logout'])) {
    session_unset();
    session_destroy();
    header("Location: vulnerable_script.php");
    exit;
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Enterprise Web App (Patched)</title>
</head>
<body>
    <h1>Enterprise Web App (Patched)</h1>

    <?php if (!isset($_SESSION['user_id'])): ?>
        <h2>Login</h2>
        <?php if (isset($error)) echo "<p style='color:red'>" . sanitize_output($error) . "</p>"; ?>
        <?php if (isset($success)) echo "<p style='color:green'>" . sanitize_output($success) . "</p>"; ?>
        <form method="POST">
            <input type="text" name="username" placeholder="Username" required><br>
            <input type="password" name="password" placeholder="Password" required><br>
            <button type="submit" name="login">Login</button>
        </form>

        <h2>Register</h2>
        <form method="POST">
            <input type="text" name="username" placeholder="Username" required><br>
            <input type="email" name="email" placeholder="Email" required><br>
            <input type="password" name="password" placeholder="Password" required><br>
            <button type="submit" name="register">Register</button>
        </form>
    <?php else: ?>
        <h2>Welcome, <?php echo sanitize_output($_SESSION['username']); ?></h2>
        <?php if (isset($error)) echo "<p style='color:red'>" . sanitize_output($error) . "</p>"; ?>
        <?php if (isset($success)) echo "<p style='color:green'>" . sanitize_output($success) . "</p>"; ?>

        <h3>Update Profile</h3>
        <form method="POST">
            <textarea name="bio" placeholder="Bio"></textarea><br>
            <input type="text" name="website" placeholder="Website"><br>
            <button type="submit" name="update_profile">Update Profile</button>
        </form>

        <h3>Upload File</h3>
        <form method="POST" enctype="multipart/form-data">
            <input type="file" name="user_file" required><br>
            <button type="submit">Upload</button>
        </form>

        <h3>Search Products</h3>
        <form method="GET">
            <input type="text" name="search" value="<?php echo sanitize_output($_GET['search'] ?? ''); ?>">
            <button type="submit">Search</button>
        </form>
        <?php if (!empty($search_results)): ?>
            <ul>
            <?php foreach ($search_results as $row): ?>
                <li><?php echo sanitize_output($row['name']); ?> - $<?php echo sanitize_output($row['price']); ?></li>
            <?php endforeach; ?>
            </ul>
        <?php endif; ?>

        <?php if ($_SESSION['role'] === 'admin'): ?>
            <h3>Admin Panel</h3>
            <a href="?admin_action=list_users">List Users</a> | 
            <a href="?admin_action=view_logs&log_file=app.log">View Logs</a>

            <?php if (!empty($users)): ?>
                <ul>
                <?php foreach ($users as $u): ?>
                    <li><?php echo sanitize_output($u['username']); ?> (<?php echo sanitize_output($u['email']); ?>)</li>
                <?php endforeach; ?>
                </ul>
            <?php endif; ?>

            <?php if (!empty($log_content)): ?>
                <pre><?php echo sanitize_output($log_content); ?></pre>
            <?php endif; ?>
        <?php endif; ?>

        <h3>Password Reset</h3>
        <form method="POST">
            <input type="email" name="email" placeholder="Enter your email">
            <button type="submit" name="reset_password">Reset Password</button>
        </form>

        <h3>XML Input</h3>
        <form method="POST">
            <textarea name="xml_data" placeholder="<root>...</root>"></textarea><br>
            <button type="submit">Submit XML</button>
        </form>

        <p><a href="?logout=1">Logout</a></p>
    <?php endif; ?>
</body>
</html>

File: vulnerable_script.py
#!/usr/bin/env python3
"""
Patched vulnerable_script.py — secure Flask app keeping original features.

Security fixes:
- Input validation & sanitization
- Strong password hashing (werkzeug)
- Session fixation protection (session.clear())
- Removed insecure pickle, replaced with JSON
- Replaced xml.etree with defusedxml (prevents XXE)
- Removed shell=True execution; implemented command whitelist
- Secure file upload (secure_filename, extension whitelist, size limit)
- Log safely via logger, not raw writes
- Environment-based secrets (no hardcoded secret keys)
"""

import os
import re
import json
import secrets
import sqlite3
import hmac
import logging
import subprocess
from datetime import datetime, timedelta
from pathlib import Path
from flask import (
    Flask, request, session, redirect, render_template_string,
    make_response, jsonify, send_file, abort
)
from werkzeug.security import generate_password_hash, check_password_hash
from werkzeug.utils import secure_filename
from markupsafe import escape
from defusedxml.ElementTree import fromstring as safe_fromstring

# ------------------- App config -------------------
app = Flask(__name__)

# Use env var for secret key; fallback to strong random (do NOT hardcode)
app.secret_key = os.environ.get("APP_SECRET_KEY") or secrets.token_hex(32)

BASE_DIR = Path.home() / "vuln_lab"
UPLOAD_DIR = BASE_DIR / "uploads"
LOG_DIR = BASE_DIR / "logs"
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)
LOG_DIR.mkdir(parents=True, exist_ok=True)

app.config['UPLOAD_FOLDER'] = str(UPLOAD_DIR) + os.sep
app.config['LOG_PATH'] = str(LOG_DIR) + os.sep
app.config['MAX_CONTENT_LENGTH'] = 5 * 1024 * 1024  # 5 MB

# Configure logger (do not log raw user input)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s %(levelname)s %(message)s',
    handlers=[
        logging.FileHandler(os.path.join(app.config['LOG_PATH'], 'app.log')),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

# ------------------- DB helpers -------------------
DB_PATH = BASE_DIR / "enterprise.db"

def get_db_connection():
    conn = sqlite3.connect(str(DB_PATH), detect_types=sqlite3.PARSE_DECLTYPES)
    conn.row_factory = sqlite3.Row
    return conn

def init_database():
    db = get_db_connection()
    db.executescript("""
    CREATE TABLE IF NOT EXISTS users (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      username TEXT UNIQUE NOT NULL,
      password TEXT NOT NULL,
      email TEXT UNIQUE NOT NULL,
      role TEXT DEFAULT 'user',
      reset_token TEXT,
      reset_expires TIMESTAMP,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    );
    CREATE TABLE IF NOT EXISTS user_profiles (
      user_id INTEGER PRIMARY KEY,
      bio TEXT,
      website TEXT
    );
    CREATE TABLE IF NOT EXISTS products (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      user_id INTEGER,
      name TEXT NOT NULL,
      price REAL,
      description TEXT
    );
    CREATE TABLE IF NOT EXISTS user_files (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      user_id INTEGER,
      original_name TEXT,
      stored_name TEXT,
      uploaded_at DATETIME DEFAULT CURRENT_TIMESTAMP
    );
    """)
    db.commit()
    db.close()

init_database()

# ------------------- Utilities -------------------
USERNAME_RE = re.compile(r'^[A-Za-z0-9_.-]{3,30}$')
EMAIL_RE = re.compile(r'^[^@]+@[^@]+\.[^@]+$')
ALLOWED_EXT = {'jpg','jpeg','png','pdf','txt'}

def valid_username(u):
    return bool(USERNAME_RE.match(u or ""))

def valid_email(e):
    return bool(EMAIL_RE.match(e or ""))

def sanitize_text(s, maxlen=2000):
    if s is None:
        return ""
    return str(s).strip()[:maxlen]

def allowed_file(filename):
    ext = filename.rsplit('.',1)[-1].lower() if '.' in filename else ''
    return ext in ALLOWED_EXT

def hash_password(pw):
    return generate_password_hash(pw)

def verify_password(pw, hashed):
    return check_password_hash(hashed, pw)

# ------------------- Views -------------------
HOME_HTML = """<!doctype html>
<title>Patched Python App</title>
<h1>Patched Python App</h1>
{% if not session.get('user_id') %}
<form method=post action="/register">
  <h3>Register</h3>
  <input name=username placeholder="username" required><br>
  <input name=email placeholder="email" required><br>
  <input name=password type=password placeholder="password" required><br>
  <button>Register</button>
</form>

<form method=post action="/login">
  <h3>Login</h3>
  <input name=username placeholder="username" required><br>
  <input name=password type=password placeholder="password" required><br>
  <button>Login</button>
</form>
{% else %}
<p>Welcome {{ session.get('username') }} ({{ session.get('role') }})</p>

<form method=post action="/update_profile">
  <h4>Profile</h4>
  <textarea name=bio rows=3 placeholder="bio"></textarea><br>
  <input name=website placeholder="website"><br>
  <button>Update Profile</button>
</form>

<form method=post action="/upload" enctype=multipart/form-data>
  <h4>Upload</h4>
  <input type=file name=user_file required><br>
  <button>Upload</button>
</form>

<form method=get action="/search">
  <h4>Search</h4>
  <input name=search placeholder="search term"><button>Search</button>
</form>

<form method=post action="/set_preference">
  <h4>Preference</h4>
  <input name=preference placeholder="preference">
  <button>Set</button>
</form>

<form method=post action="/reset_password">
  <h4>Reset</h4>
  <input name=email placeholder="email"><button>Reset</button>
</form>

{% if session.get('role') == 'admin' %}
  <p><a href="/admin?action=list_users">Admin: list users</a></p>
{% endif %}

<p><a href="/logout">Logout</a></p>
{% endif %}
"""

@app.route("/", methods=["GET"])
def index():
    name = request.args.get("name", "Guest")
    # escape user input to avoid XSS
    safe_name = escape(name)
    return render_template_string(HOME_HTML, session=session, safe_name=safe_name)

# ----- Register -----
@app.route("/register", methods=["POST"])
def register():
    username = request.form.get("username", "").strip()
    email = request.form.get("email", "").strip()
    password = request.form.get("password", "")

    if not valid_username(username) or not valid_email(email) or len(password) < 8:
        return "Invalid registration data", 400

    hashed = hash_password(password)
    db = get_db_connection()
    try:
        db.execute("INSERT INTO users (username, password, email) VALUES (?, ?, ?)", (username, hashed, email))
        db.commit()
    except Exception:
        logger.info("Registration failed for %s", username)
        db.close()
        return "Registration failed", 400
    # prevent session fixation
    session.clear()
    row = db.execute("SELECT id FROM users WHERE username=?", (username,)).fetchone()
    session['user_id'] = row['id']
    session['username'] = username
    session['role'] = 'user'
    db.close()
    return redirect("/")

# ----- Login -----
@app.route("/login", methods=["POST"])
def login():
    username = request.form.get("username", "").strip()
    password = request.form.get("password", "")

    db = get_db_connection()
    row = db.execute("SELECT id, username, role, password FROM users WHERE username=?", (username,)).fetchone()
    db.close()
    if not row or not verify_password(password, row["password"]):
        logger.info("Failed login for %s", username)
        return "Invalid credentials", 401
    session.clear()
    session['user_id'] = row['id']
    session['username'] = row['username']
    session['role'] = row['role']
    return redirect("/")

# ----- Update Profile -----
@app.route("/update_profile", methods=["POST"])
def update_profile():
    if not session.get('user_id'):
        return "Unauthorized", 401
    bio = sanitize_text(request.form.get("bio", "")) 
    website = sanitize_text(request.form.get("website", ""))
    db = get_db_connection()
    db.execute("INSERT OR REPLACE INTO user_profiles (user_id, bio, website) VALUES (?, ?, ?)", (session['user_id'], bio, website))
    db.commit()
    db.close()
    return redirect("/")

# ----- Upload -----
@app.route("/upload", methods=["POST"])
def upload():
    if not session.get('user_id'):
        return "Unauthorized", 401
    if 'user_file' not in request.files:
        return "No file", 400
    f = request.files['user_file']
    filename = secure_filename(f.filename)
    if filename == "":
        return "Invalid filename", 400
    if not allowed_file(filename):
        return "Invalid file type", 400
    # random stored name
    ext = filename.rsplit('.',1)[-1].lower()
    stored = secrets.token_hex(12) + "." + ext
    path = UPLOAD_DIR / stored
    f.save(path)
    os.chmod(path, 0o640)
    db = get_db_connection()
    db.execute("INSERT INTO user_files (user_id, original_name, stored_name) VALUES (?, ?, ?)", (session['user_id'], filename, stored))
    db.commit()
    db.close()
    logger.info("User %s uploaded a file", session.get('username'))
    return redirect("/")

# ----- Search (parameterized) -----
@app.route("/search", methods=["GET"])
def search():
    term = sanitize_text(request.args.get("search", ""))
    db = get_db_connection()
    rows = db.execute("SELECT name, price FROM products WHERE name LIKE ?", (f"%{term}%",)).fetchall()
    db.close()
    out = "<h3>Results</h3><ul>"
    for r in rows:
        out += f"<li>{escape(r['name'])} - ${escape(str(r['price']))}</li>"
    out += "</ul><p><a href='/'>Back</a></p>"
    return out

# ----- Admin -----
@app.route("/admin", methods=["GET"])
def admin():
    if session.get('role') != 'admin':
        return "Unauthorized", 403
    action = request.args.get('action', '')
    if action == 'list_users':
        db = get_db_connection()
        rows = db.execute("SELECT id, username, email, role FROM users").fetchall()
        db.close()
        return "<pre>" + "\n".join([f"{r['id']}: {escape(r['username'])} ({escape(r['email'])}) role={escape(r['role'])}" for r in rows]) + "</pre>"
    if action == 'exec':
        # whitelist commands — do NOT use shell=True
        cmd_key = request.args.get('cmd', '')
        allowed = {
            'whoami': ['whoami'],
            'uptime': ['uptime'],
            'date': ['date']
        }
        if cmd_key not in allowed:
            return "Command not allowed", 400
        try:
            result = subprocess.check_output(allowed[cmd_key], text=True)
            return "<pre>" + escape(result) + "</pre>"
        except Exception:
            logger.exception("Admin command failed")
            return "Command failed", 500
    return "Admin panel"

# ----- Reset Password -----
@app.route("/reset_password", methods=["POST"])
def reset_password():
    email = request.form.get('email', '').strip()
    if not valid_email(email):
        return "Invalid email", 400
    token = secrets.token_urlsafe(32)
    expires = (datetime.utcnow() + timedelta(hours=1)).isoformat()
    db = get_db_connection()
    db.execute("UPDATE users SET reset_token=?, reset_expires=? WHERE email=?", (token, expires, email))
    db.commit()
    db.close()
    logger.info("Reset token for %s: %s", email, token)
    return "If the email exists, a reset link was issued (logged)."

# ----- XML processing (defusedxml prevents XXE) -----
@app.route("/xml", methods=["POST"])
def xml_process():
    xml_data = request.form.get('xml_data', '')
    try:
        root = safe_fromstring(xml_data)
        return "Processed XML root: " + escape(root.tag)
    except Exception:
        logger.info("Invalid XML submitted")
        return "Invalid XML", 400

# ----- Download file (owner or admin only) -----
@app.route("/download/<int:file_id>", methods=["GET"])
def download(file_id):
    db = get_db_connection()
    row = db.execute("SELECT * FROM user_files WHERE id=?", (file_id,)).fetchone()
    db.close()
    if not row:
        return "Not found", 404
    if session.get('user_id') != row['user_id'] and session.get('role') != 'admin':
        return "Unauthorized", 403
    path = UPLOAD_DIR / row['stored_name']
    if not path.exists():
        return "Missing", 404
    return send_file(path, as_attachment=True, download_name=row['original_name'])

# ----- Preference cookie -----
@app.route("/set_preference", methods=["POST"])
def set_preference():
    pref = sanitize_text(request.form.get('preference',''), maxlen=100)
    resp = make_response(redirect("/"))
    resp.set_cookie("user_preference", pref, max_age=365*24*3600, httponly=True, samesite='Lax')
    return resp

# ----- Logout -----
@app.route("/logout")
def logout():
    session.clear()
    return redirect("/")

# ----- Error handlers -----
@app.errorhandler(500)
def internal_error(e):
    logger.exception("Internal server error")
    return "Internal server error", 500

@app.errorhandler(413)
def payload_too_large(e):
    return "File too large", 413

# ----- Run -----
if __name__ == "__main__":
    # Run on localhost for lab testing
    app.run(host="127.0.0.1", port=5000, debug=False)
